apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts

```diff
@@ -0,0 +1,90 @@
+
+import {
+  ControlValuesEntity,
+  ControlValuesRepository,
+  NotificationStepEntity,
+  NotificationTemplateRepository,
+} from '@novu/dal';
+import { ControlVariablesLevelEnum } from '@novu/shared';
+import { GetPreferences, GetPreferencesCommand, GetPreferencesResponseDto } from '@novu/application-generic';
+
+import { GetWorkflowCommand } from './get-workflow.command';
+import { NotificationTemplateMapper } from '../../mappers/notification-template-mapper';
+import { WorkflowNotFoundException } from '../../exceptions/workflow-not-found-exception';
+import { WorkflowResponseDto } from '../../dto/workflow-response-dto';
+
+@Injectable()
+export class GetWorkflowUseCase {
+  constructor(
+    private notificationTemplateRepository: NotificationTemplateRepository,
+    private controlValuesRepository: ControlValuesRepository,
+    private getPreferencesUseCase: GetPreferences
+  ) {}
+  async execute(command: GetWorkflowCommand): Promise<WorkflowResponseDto> {
+    const notificationTemplateEntity = await this.notificationTemplateRepository.findByIdQuery({
+      id: command._workflowId,
+      environmentId: command.user.environmentId,
+    });
+
+    if (!notificationTemplateEntity) {
+      throw new WorkflowNotFoundException(command._workflowId);
+    }
+    const stepIdToControlValuesMap = await this.getControlsValuesMap(notificationTemplateEntity.steps, command);
+    const preferences = await this.getPreferencesForWorkflow(command);
+
+    return NotificationTemplateMapper.toResponseWorkflowDto(
+      notificationTemplateEntity,
+      preferences,
+      stepIdToControlValuesMap
+    );
+  }
+
+  private async getPreferencesForWorkflow(
+    getWorkflowCommand: GetWorkflowCommand
+  ): Promise<GetPreferencesResponseDto | undefined> {
+    const command = {
+      environmentId: getWorkflowCommand.user.environmentId,
+      organizationId: getWorkflowCommand.user.organizationId,
+      templateId: getWorkflowCommand._workflowId,
+    } as GetPreferencesCommand;
+
+    try {
+      const workflowPreferences = await this.getPreferencesUseCase.execute(GetPreferencesCommand.create(command));
+
+      return workflowPreferences;
+    } catch (e) {
+      Logger.error('Error getting preferences for workflow', e);
+
+      return undefined;
+    }
+  }
+
+  private async getControlsValuesMap(
+    steps: NotificationStepEntity[],
+    command: GetWorkflowCommand
+  ): Promise<{ [key: string]: ControlValuesEntity }> {
+    const acc: { [key: string]: ControlValuesEntity } = {};
+
+    for (const step of steps) {
+      const controlValuesEntity = await this.buildControlValuesForStep(step, command);
+      if (controlValuesEntity) {
+        acc[step._templateId] = controlValuesEntity;
+      }
+    }
+
+    return acc;
+  }
+  private async buildControlValuesForStep(
+    step: NotificationStepEntity,
+    command: GetWorkflowCommand
+  ): Promise<ControlValuesEntity | null> {
+    return await this.controlValuesRepository.findFirst({
```

Comment:
As a follow-up question, since we have the workflow_id, why do we need the organization and environment IDs in the query? We can fetch them using a simpler workflow_id and step_id index.

Reply:
just find first returns a single object vs a potential array so i prefer to keep it